File: src/openclaw_predicate_provider/openclaw_hooks.py
"""Thin OpenClaw hook interface for guarded tool execution."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Protocol

from .adapter import ToolAdapter

ToolExecutor = Callable[[dict[str, Any]], Awaitable[Any]]


class OpenClawHookContext(Protocol):
    """Protocol for hook context expected from OpenClaw runtime."""

    session_id: str
    source: str
    tenant_id: str | None
    user_id: str | None
    trace_id: str | None


@dataclass(slots=True)
class HookEnvelope:
    """Normalized hook payload for adapter consumption."""

    tool_name: str
    args: dict[str, Any]
    session_id: str
    source: str
    tenant_id: str | None = None
    user_id: str | None = None
    trace_id: str | None = None

    def context(self) -> dict[str, Any]:
        """Render policy context claims sent to Predicate."""
        return {
            "session_id": self.session_id,
            "source": self.source,
            "tenant_id": self.tenant_id,
            "user_id": self.user_id,
            "trace_id": self.trace_id,
        }


class OpenClawHooks:
    """Hook surface matching likely OpenClaw tool callback wiring."""

    def __init__(self, adapter: ToolAdapter):
        self._adapter = adapter

    async def on_cmd_run(self, envelope: HookEnvelope, execute: ToolExecutor) -> Any:
        return await self._adapter.run_shell(
            args=envelope.args,
            context=envelope.context(),
            execute=execute,
        )

    async def on_fs_read(self, envelope: HookEnvelope, execute: ToolExecutor) -> Any:
        return await self._adapter.read_file(
            args=envelope.args,
            context=envelope.context(),
            execute=execute,
        )

    async def on_http_request(
        self,
        envelope: HookEnvelope,
        execute: ToolExecutor,
    ) -> Any:
        return await self._adapter.http_request(
            args=envelope.args,
            context=envelope.context(),
            execute=execute,
        )

File: src/openclaw-hooks.ts
import type { ToolAdapter } from "./adapter.js";

export class HookEnvelope {
  readonly toolName: string;
  readonly args: Record<string, unknown>;
  readonly sessionId: string;
  readonly source: string;
  readonly tenantId?: string;
  readonly userId?: string;
  readonly traceId?: string;

  constructor(params: {
    toolName: string;
    args: Record<string, unknown>;
    sessionId: string;
    source: string;
    tenantId?: string;
    userId?: string;
    traceId?: string;
  }) {
    this.toolName = params.toolName;
    this.args = params.args;
    this.sessionId = params.sessionId;
    this.source = params.source;
    this.tenantId = params.tenantId;
    this.userId = params.userId;
    this.traceId = params.traceId;
  }

  context(): Record<string, unknown> {
    return {
      session_id: this.sessionId,
      source: this.source,
      tenant_id: this.tenantId,
      user_id: this.userId,
      trace_id: this.traceId,
    };
  }
}

export class OpenClawHooks {
  constructor(private readonly adapter: Pick<ToolAdapter, "runShell" | "readFile" | "httpRequest">) {}

  onCmdRun(
    envelope: HookEnvelope,
    execute: (args: Record<string, unknown>) => Promise<unknown>,
  ) {
    return this.adapter.runShell({
      args: envelope.args,
      context: envelope.context(),
      execute,
    });
  }

  onFsRead(
    envelope: HookEnvelope,
    execute: (args: Record<string, unknown>) => Promise<unknown>,
  ) {
    return this.adapter.readFile({
      args: envelope.args,
      context: envelope.context(),
      execute,
    });
  }

  onHttpRequest(
    envelope: HookEnvelope,
    execute: (args: Record<string, unknown>) => Promise<unknown>,
  ) {
    return this.adapter.httpRequest({
      args: envelope.args,
      context: envelope.context(),
      execute,
    });
  }
}
